### PR TITLE
Command coordinator supports flows and events

### PR DESCRIPTION
**Pull request by masayag**
* Describes the support of command coordinator in flows and events
* Describes the refactor done in command coordinator internals

### CODE REVIEW

**Comment by masayag** [comment]
@laravot @oourfali please review.

File: source/develop/release-management/features/engine/command-coordinator-flows-and-events.html.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+*   Email: <laravot@redhat.com>
+
+### Detailed Description
+The existing CoCo implementation provides api for registering a callback to be executed for a command.
+However, the api doesn't define which component is responsible to manage the entire command’s flow. Therefore the callback
+is responsible to invoke the next step in the chain, either by managing its own state machine in `CommandCallback.doPolling()`
+or on `CommandCallback.onSuccess()` or `CommandCallback.onFailure()`.
+
+In addition, there are several mechanism within ovirt-engine to achieve the same result, i.e. SEAT and Command Coordinator.
+SEAT was a mid-term solution and once Command Coordinator was introduced, it should be deprecated and removed.
```
Comment (mykaul) [inline source/develop/release-management/features/engine/command-coordinator-flows-and-events.html.md]: was introduced -> is introduced

File: source/develop/release-management/features/engine/command-coordinator-flows-and-events.html.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+  * **CommandExecutor** is responsible for invoking commands by command-coordinator in the system.
+  * **CommandsRepository** is responsible for handling command's related information, callbacks and caching.
+  * **ConcurrentChildCommandsExecutionCallback** implements a callback which should process its child commands in parallel,
+   and continues after all of its child commands executions are completed
+  * **SerialChildCommandsExecutionCallback** implements a callback which executes its child commands serially, this
+   callback will terminate the execute if any of the child commands fails.
+
+#### Support Flows with Command Coordinator ####
+The definition of a flow is determined by the command, by implementing the **SerialChildExecutingCommand** interface which
+define the next operation to perform by:
```
Comment (mykaul) [inline source/develop/release-management/features/engine/command-coordinator-flows-and-events.html.md]: define -> defines

File: source/develop/release-management/features/engine/command-coordinator-flows-and-events.html.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+  `boolean SerialChildExecutingCommand.performNextOperation()`
+
+The command which implements this interface will define the next step of its execution. The action will be triggered by
+the commands callback (**SerialChildExecutingCommand**).
+The invoked actions via **performNextOperation()** can be either commands, commands with callbacks or tasks or other
+execution pieces.
+The orchestration of the flow will be done inside the command, and its progress will be monitored by command coordinator,
+which upon its child completion will determine if it should continued or canceled.
```
Comment (mykaul) [inline source/develop/release-management/features/engine/command-coordinator-flows-and-events.html.md]: continued -> be continued

File: source/develop/release-management/features/engine/command-coordinator-flows-and-events.html.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+The following sequence diagram describes the components and their interaction in executing a command's flow:
+ 
+![](CommandFlow.png "CommandFlow.png")
+
+The following diagram shows the supported flows, which might be a combination of a serial and concurrent commands invocation:
+
+![](oCoFlowCommandWithTasks.png "oCoFlowCommandWithTasks.png")
+
+#### Support Events with Command Coordinator ####
+Command should be able to register for an event. The registration could be done either from the command.
```
Comment (mykaul) [inline source/develop/release-management/features/engine/command-coordinator-flows-and-events.html.md]: either from the command ...    - and?

File: source/develop/release-management/features/engine/command-coordinator-flows-and-events.html.md [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+Registering the command to the event will be done prior to invoking the action which will trigger the event:
+The subscribed event will be processed asynchronously, hence it would be advisable to register prior to create the
```
Comment (mykaul) [inline source/develop/release-management/features/engine/command-coordinator-flows-and-events.html.md]: create -> creating

**Comment by mureinik** [comment]
@thatdocslady / @sandrobonazzola  Can this be merged please?

**Comment by masayag** [comment]
All comments were addressed. Ready to be merged.

**Pull request merged**